rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,519 @@
+
+### Migration Example
+
+The following shows one of the many possible ways to migrate the current Nixpkgs `lib` comments.
+
+> We managed to partially automate this effort with a [codemod](https://github.com/nix-community/docnix/tree/3c0531cb5b4c9f3e9069b73d19e6c4be8508d905/codemod)
+
+> Note: The current `nixdoc` feature 'Function arguments' uses a different mental model than this RFC. Arguments must now be explicitly documented inside of the lambda documentation.
+
+`lib/attrsets.nix (old format)`
+````nix
+/* Filter an attribute set by removing all attributes for which the
+   given predicate return false.
+
+   Example:
+     filterAttrs (n: v: n == "foo") { foo = 1; bar = 2; }
+     => { foo = 1; }
+
+   Type:
+     filterAttrs :: (String -> Any -> Bool) -> AttrSet -> AttrSet
+*/
+filterAttrs =
+  # Predicate taking an attribute name and an attribute value, which returns `true` to include the attribute or `false` to exclude the attribute.
+  pred:
+  # The attribute set to filter
+  set:
+  listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
+````
+
+->
+
+`lib/attrsets.nix (new format)`
+````nix
+/**
+  Filter an attribute set by removing all attributes for which the
+  given predicate return false.
+
+  # Example
+
+  ```nix
+  filterAttrs (n: v: n == "foo") { foo = 1; bar = 2; }
+  => { foo = 1; }
+  ```
+
+  # Type
+
+  ```
+  filterAttrs :: (String -> Any -> Bool) -> AttrSet -> AttrSet
+  ```
+
+  # Arguments
+
+  - [pred] Predicate taking an attribute name and an attribute value, which returns `true` to include the attribute, or `false` to exclude the attribute.
+  - [set] The attribute set to filter
+*/
+filterAttrs =
+  pred:
+  set:
+  listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
+````
+
+## Tooling
+
+All current rendering tooling solutions should support displaying the specified doc-comment format.
+
+Currently, at least:
+
+### Nixpkgs Manual renderer
+
+The current Nixpkgs manual needs to be adapted to this change.
+
+We expect changes in the following to be necessary:
+
+- [nixos_render_docs](https://github.com/NixOS/Nixpkgs/tree/e4082efedb483eb0478c3f014fa851449bca43f9/pkgs/tools/nix/nixos-render-docs/src) 
+- [nixdoc](https://github.com/nix-community/nixdoc)
+
+## Native support in Nix
+
+- `NixOS/nix` should implement native support for doc-comments.
+
+Primarily because the *Nix* language requires access to the actual evaluator for building correct relations between doc-comment and the node they document.
+
+> Note: We considered implementation details, but specifying those is out of the scope of this RFC.
+
+Related POCs:
+
+- https://github.com/NixOS/nix/pull/9054
+- https://github.com/NixOS/nix/pull/1652
+- https://github.com/NixOS/nix/pull/5527
+- https://github.com/NixOS/nix/commit/8e252320f52426c734883709eb398d2161c3fe82
+
```

Comment:
```suggestion
```

This needs its own RFC.  What is currently in this RFC is nowhere near well-described enough to "count" as an RFC, and an RFC can't delegate the RFC process' authority to bring ["new features into the Nix language"](https://github.com/hsjobeki/rfcs/blob/2987912e0f33b73b37250817953bf62fd5594a25/rfcs/0001-rfc-process.md?plain=1#L12C57-L13C63).

I have no other objections to this RFC, but slipping in this "oh yeah and we will expose this to the evaluator" without any details really doesn't fly with me.  My main criterion there is ensuring that this feature can be turned off, *and is in fact turned off* in any evaluation session which creates derivations.  As the RFC now stands I can't tell if that's the case.

Reply:
This is only in the future work section, it's not part of the RFC's design

Reply:
Then there should be no problem removing it.

I don't see the point of putting something that isn't part of an RFC in the RFC.

Reply:
+1 for removing.

Reply:
Ok Agreed. I'll remove this statement.